core/network/fabric-interop-cc/contracts/interop/membership.go

```diff
@@ -9,45 +9,334 @@
+	if err != nil {
+		return fmt.Errorf("Marshal error: %s", err)
+	}
+	return ctx.GetStub().PutState(membershipLocalKey, membershipBytes)
+}
+
+// UpdateLocalMembership cc is used to update the existing local security domain's Membership in the ledger
+func (s *SmartContract) UpdateLocalMembership(ctx contractapi.TransactionContextInterface, membershipJSON string) error {
+	// Check if the caller has network admin privileges
+	if isAdmin, err := isClientNetworkAdmin(ctx); err != nil {
+		return fmt.Errorf("Admin client check error: %+v", err)
+	} else if !isAdmin {
+		return fmt.Errorf("Caller not a network admin; access denied")
+	}
+
+	membership, err := decodeMembership([]byte(membershipJSON))
+	if err != nil {
+		return fmt.Errorf("Unmarshal error: %s", err)
+	}
+
+	membershipLocalKey, err := ctx.GetStub().CreateCompositeKey(membershipObjectType, []string{membershipLocalSecurityDomain})
+	_, getErr := s.GetMembershipBySecurityDomain(ctx, membershipLocalSecurityDomain)
+	if getErr != nil {
+		return getErr
+	}
+
+	// Check if certificates chains in this membership record are valid
+	err = validateMemberCertChains(membership)
+	if err != nil {
+		return err
+	}
+
+	membershipBytes, err := json.Marshal(membership)
+	if err != nil {
+		return fmt.Errorf("Marshal error: %s", err)
+	}
+	return ctx.GetStub().PutState(membershipLocalKey, membershipBytes)
+
+}
+
 // CreateMembership cc is used to store a Membership in the ledger
-func (s *SmartContract) CreateMembership(ctx contractapi.TransactionContextInterface, membershipJSON string) error {
+// TODO: Remove call to 'createMembership' after creating Corda IIN Agents.
+func (s *SmartContract) CreateMembership(ctx contractapi.TransactionContextInterface, counterAttestedMembershipSerialized string) error {
+	// Check if the caller has IIN agent privileges
+	if isIINAgent, err := isClientIINAgent(ctx); err != nil {
+		return fmt.Errorf("IIN Agent client check error: %s", err)
+	} else if !isIINAgent {
+		// Check if the caller has network admin privileges
+		if isAdmin, err := isClientNetworkAdmin(ctx); err != nil {
+			return fmt.Errorf("Admin client check error: %s", err)
+		} else if !isAdmin {
+			return fmt.Errorf("Caller neither a network admin nor an IIN Agent; access denied")
+		}
+		return s.createMembership(ctx, counterAttestedMembershipSerialized)		// HACK to handle unattested memberships (for Corda) for backward compatibility
+	}
+
+	// Check if caller is one of the authorized IIN agents
+	membershipLocal, err := s.GetMembershipBySecurityDomain(ctx, membershipLocalSecurityDomain)
+	if err != nil {
+		return err
+	}
+	membershipLocalDecoded, err := decodeMembership([]byte(membershipLocal))
+	if err != nil {
+		return fmt.Errorf("Local Membership JSON Unmarshal error: %s", err)
+	}
+	callerId := ctx.GetClientIdentity()
+	callerCert, err := callerId.GetX509Certificate()
+	if err != nil {
+		return err
+	}
+	err = checkUserInMembership(callerCert, membershipLocalDecoded)
```

Comment:
What's the difference in using this function, and using one already available `verifyMemberInSecurityDomain`?
Instead of fetching membership in line 196, instead let the existing function fetch it using the key. Also existing function some additional safety checks.

Reply:
The only difference I can see is the existing function takes cert as string, while new one takes isntance of x509Certificate class. So can we unify the two functions?

Reply:
You are right. I've removed `checkUserInMembership` and refactored `verifyMemberInSecurityDomain` and the code calling it appropriately.